pkg/rtc/types/interfaces.go

```diff
@@ -32,86 +32,110 @@ const (
-	HandleAnswer(sdp webrtc.SessionDescription) error
-	AddICECandidate(candidate webrtc.ICECandidateInit, target livekit.SignalTarget) error
-	AddSubscriber(op Participant, params AddSubscriberParams) (int, error)
-	RemoveSubscriber(op Participant, trackID livekit.TrackID)
+
+	// returns list of participant identities that the current participant is subscribed to
+	GetSubscribedParticipants() []livekit.ParticipantID
+
+	GetAudioLevel() (level uint8, active bool)
+	GetConnectionQuality() *livekit.ConnectionQualityInfo
+
+	// server sent messages
 	SendJoinResponse(info *livekit.Room, otherParticipants []*livekit.ParticipantInfo, iceServers []*livekit.ICEServer) error
 	SendParticipantUpdate(participants []*livekit.ParticipantInfo, updatedAt time.Time) error
 	SendSpeakerUpdate(speakers []*livekit.SpeakerInfo) error
 	SendDataPacket(packet *livekit.DataPacket) error
 	SendRoomUpdate(room *livekit.Room) error
 	SendConnectionQualityUpdate(update *livekit.ConnectionQualityUpdate) error
-	SetTrackMuted(trackID livekit.TrackID, muted bool, fromAdmin bool)
-	GetAudioLevel() (level uint8, active bool)
-	GetConnectionQuality() *livekit.ConnectionQualityInfo
-	IsSubscribedTo(participantID livekit.ParticipantID) bool
-	// returns list of participant identities that the current participant is subscribed to
-	GetSubscribedParticipants() []livekit.ParticipantID
+	SubscriptionPermissionUpdate(publisherID livekit.ParticipantID, trackID livekit.TrackID, allowed bool)
+
+	// callbacks
+	OnStateChange(func(p Participant, oldState livekit.ParticipantInfo_State))
+	// OnTrackUpdated - one of its publishedTracks changed in status
+	OnTrackUpdated(callback func(Participant, PublishedTrack))
+	OnMetadataUpdate(callback func(Participant))
+	OnDataPacket(callback func(Participant, *livekit.DataPacket))
+	OnClose(_callback func(Participant, map[livekit.TrackID]livekit.ParticipantID))
+
+	// updates from remotes
+	UpdateSubscribedQuality(nodeID string, trackID livekit.TrackID, maxQuality livekit.VideoQuality) error
+	UpdateMediaLoss(nodeID string, trackID livekit.TrackID, fractionalLoss uint32) error
+
+	// session migration
+	SetMigrateState(s MigrateState)
+	MigrateState() MigrateState
+	AddMigratedTrack(cid string, ti *livekit.TrackInfo)
+	SetPreviousAnswer(previous *webrtc.SessionDescription)
+}
+
+//counterfeiter:generate . Participant
+type Participant interface {
+	ID() livekit.ParticipantID
+	Identity() livekit.ParticipantIdentity
+
+	ToProto() *livekit.ParticipantInfo
+
+	SetMetadata(metadata string)
+
+	GetPublishedTrack(sid livekit.TrackID) PublishedTrack
+	GetPublishedTracks() []PublishedTrack
+
+	AddSubscriber(op Participant, params AddSubscriberParams) (int, error)
+	RemoveSubscriber(op Participant, trackID livekit.TrackID)
 
 	// permissions
-	CanPublish() bool
-	CanSubscribe() bool
-	CanPublishData() bool
 	Hidden() bool
-	IsRecorder() bool
-	SubscriberAsPrimary() bool
 
 	Start()
```

Comment:
`Start()` and `DebugInfo()` below is not currently needed in `RemoteParticipant`, but left it in there as they seem useful and also for symmetry with `Close` method.